WoodStar/WoodStar/WoodStarConnection.cs

```diff
@@ -0,0 +1,157 @@
+
+            public override void Flush() => _stream.Flush();
+
+            public override Task FlushAsync(CancellationToken cancellationToken) => _stream.FlushAsync(cancellationToken);
+
+            public override int Read(byte[] buffer, int offset, int count) => _stream.Read(buffer, offset, count);
+
+            public override long Seek(long offset, SeekOrigin origin) => _stream.Seek(offset, origin);
+
+            public override void SetLength(long value) => _stream.SetLength(value);
+
+            public override void Write(byte[] buffer, int offset, int count) => _stream.Write(buffer, offset, count);
+
+            public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+            {
+                if (_intercept)
+                {
+                    if (_packetBytes == 0)
+                    {
+                        // Process header
+                        var header = new byte[TdsHeader.HeaderSize];
+                        var headerBytes = 0;
+                        do
+                        {
+                            var headerReadCount = await _stream.ReadAsync(header.AsMemory()[headerBytes..TdsHeader.HeaderSize], cancellationToken);
+                            if (headerReadCount == 0)
+                            {
+                                throw new EndOfStreamException();
+                            }
+                            headerBytes += headerReadCount;
+                        } while (headerBytes < TdsHeader.HeaderSize);
+
+                        _packetBytes = 256 * header[2] + header[3] - TdsHeader.HeaderSize;
+                    }
+
+                    var readCount = await _stream.ReadAsync(buffer.Slice(0, Math.Min(buffer.Length, _packetBytes)), cancellationToken);
+                    if (readCount == 0)
+                    {
+                        throw new EndOfStreamException();
+                    }
+
+                    _packetBytes -= readCount;
+                    return readCount;
+                }
+
+                return await _stream.ReadAsync(buffer, cancellationToken);
+            }
+
+            public override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+            {
+                if (_intercept)
+                {
+                    var modifiedLength = buffer.Length + TdsHeader.HeaderSize;
+                    var modifiedBuffer = ArrayPool<byte>.Shared.Rent(modifiedLength);
+                    buffer.CopyTo(modifiedBuffer.AsMemory(TdsHeader.HeaderSize));
+
+                    var tdsHeader = new TdsHeader(PacketType.PreLogin, PacketStatus.EOM, modifiedLength, 0, 1);
+                    tdsHeader.WriteToBuffer(modifiedBuffer);
+                    await _stream.WriteAsync(modifiedBuffer.AsMemory().Slice(0, modifiedLength), cancellationToken);
+                    ArrayPool<byte>.Shared.Return(modifiedBuffer);
```

Comment:
I wonder whether it makes sense to clear the buffer before returning it to the pool, since we do send password and other tls-related stuff.